jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java

```diff
@@ -126,35 +126,11 @@ public static MultiPartFormData.Parts getParts(Content.Source content, Attribute
      * @param attributes the attributes where the futureParts are tracked.
      * @param contentType the value of the {@link HttpHeader#CONTENT_TYPE} header.
      * @param config the multipart configuration.
-     * @param immediate The action to take if the FormFields are available immediately (from within the scope of the call to this method).
-     * @param future The action to take when the FormFields are available, if they are not available immediately.  The {@link org.eclipse.jetty.util.thread.Invocable.InvocationType}
-     *               of this parameter will be used as the type for any implementation calls to {@link Content.Source#demand(Runnable)}.
+     * @param promise The action to take when the {@link Parts} are available.
      */
-    public static void onParts(Content.Source content, Attributes attributes, String contentType, MultiPartConfig config, Promise<Parts> immediate, Promise.Invocable<Parts> future)
+    public static void onParts(Content.Source content, Attributes attributes, String contentType, MultiPartConfig config, Promise.Invocable<Parts> promise)
     {
-        CompletableFuture<Parts> futureParts = from(content, future.getInvocationType(), attributes, contentType, config);
-
-        if (futureParts.isDone())
-        {
-            Parts parts = null;
-            Throwable error = null;
-            try
-            {
-                parts = futureParts.get();
-            }
-            catch (Throwable t)
-            {
-                error = t;
-            }
-            if (error != null)
-                immediate.failed(error);
-            else
-                immediate.succeeded(parts);
-        }
-        else
-        {
-            futureParts.whenComplete(future);
-        }
+        from(content, promise.getInvocationType(), attributes, contentType, config).whenComplete(promise);
```

Comment:
This can be written the other way around, leveraging existing methods that have been introduced a while ago.

```suggestion
        promise.completeWith(from(content, promise.getInvocationType(), attributes, contentType, config));
```

In this way, we can simplify `Promise.Invocable` and eventually just make `Promise` implement `Invocable` without the need of an extra class.

Reply:
However, `promise.completeWith` is an extra allocation as it creates a lambda on every call.    Having a Promise that is also a BiConsumer is a reasonable optimization that avoids this allocation.